File: embulk-core/src/main/java/org/embulk/spi/EncoderPlugin.java
package org.embulk.spi;

import org.embulk.config.ConfigSource;
import org.embulk.config.TaskSource;

public interface EncoderPlugin {
    interface Control {
        void run(TaskSource taskSource);
    }

    void transaction(ConfigSource config, EncoderPlugin.Control control);

    FileOutput open(TaskSource taskSource, FileOutput fileOutput);
}

File: lib/embulk/encoder_plugin.rb
module Embulk

  require 'embulk/data_source'

  class EncoderPlugin
    def self.transaction(config, &control)
      raise NotImplementedError, "EncoderPlugin.transaction(config, &control) must be implemented"
    end

    # TODO

    # TODO new_java

    def self.from_java(java_class)
      JavaPlugin.ruby_adapter_class(java_class, EncoderPlugin, RubyAdapter)
    end

    module RubyAdapter
      module ClassMethods
      end
      # TODO
    end
  end

end
